pkg/util/lazy/lazy.go

```diff
@@ -0,0 +1,62 @@
+// Copyright 2023 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lazy
+
+import (
+	"sync"
+	"sync/atomic"
+)
+
+// Lazy defers the evaluation of getter until it's accessed the first access.
+type Lazy[T any] interface {
+	// Get returns the value, evaluate it if necessary.
+	Get() T
+	// Evaluated returns whether the value has been evaluated or not.
+	Evaluated() bool
+}
+
+type lazy[T any] struct {
+	getter func() T
+	// res is the cached result.
+	res  T
+	done uint32
```

Comment:
I may be missing something, but since we are using a mutex anyway, it's unclear to me what the benefit of using `atomic` functions is, compared to using the mutex "everywhere":
```go
func (l *lazy[T]) Get() T {
        l.m.Lock()
	defer l.m.Unlock()
	if !l.done {
	        l.res = l.getter()
                l.done = true
	}
	return l.res
}

func (l *lazy[T]) Evaluated() bool {
        l.m.Lock()
	defer l.m.Unlock()
        return l.done
}
```

The main difference is is in the behavior of `Evaluated` if the object is currently being evaluated (`Evaluated` will now block and return true). I don't know if one approach is more desired here.

I guess another minor difference is that later calls to `Get` and `Evaluated` may be slightly faster (no need to acquire the mutex) once slow initialization has been done.

Reply:
I refered to `sync.Once`: https://github.com/golang/go/blob/go1.21.4/src/sync/once.go
According to the benchmark below, using atomic is about 10x faster than acquing a lock:
```
func (l *lazy[T]) Get() T {
	if atomic.LoadUint32(&l.done) == 0 {
		return l.doSlow()
	}
	return l.res
}

func (l *lazy[T]) GetWithLock() T {
	l.m.Lock()
	defer l.m.Unlock()
	if l.done == 0 {
		l.res = l.getter()
		l.done = 1
	}
	return l.res
}

func BenchmarkGet(b *testing.B) {
	lazyFoo := New(func() *foo {
		return &foo{}
	})
	for i := 0; i < b.N; i++ {
		lazyFoo.Get()
	}
}

func BenchmarkGetWithLock(b *testing.B) {
	lazyFoo := New(func() *foo {
		return &foo{}
	})
	for i := 0; i < b.N; i++ {
		lazyFoo.GetWithLock()
	}
}

goos: linux
goarch: amd64
pkg: antrea.io/antrea/pkg/util/lazy
cpu: Intel(R) Xeon(R) Silver 4114 CPU @ 2.20GHz
BenchmarkGet-4                  425120371                2.817 ns/op
BenchmarkGetWithLock-4          44733890                26.91 ns/op
```